dag.md

```diff
@@ -0,0 +1,234 @@
+
+**Phase two (initialization)**
+
+Adding new elements to the main instance requires their initialization the same way we prepare the main instance to receive new answers the first time. `Set<QuickTriggerable> qtSet2 = initializeTriggerables(mainInstance, evalContext, createRef, new HashSet<>(0));` starts the evaluation of all triggerables targetting a descendant of the repeat group's reference. 
+
+This is done to prepare the new elements in case they have computed values. 
+
+**Phase three (children)**
+
+Now the children elements in the new repeat group instance have been created and initialized, which means that we should evaluate any triggerable triggered by their references. `evaluateChildrenTriggerables(mainInstance, evalContext, createdElement, true, alreadyEvaluated);` starts the chain of evaluations of these triggerables as if a value would have changed in each of them. 
+
+#### c - A repeat instance is deleted
+
+`FormEntryController.deleteRepeat(...)` is called when a user deletes a repeat group. After removing the corresponding elements from the main instance, `TriggerableDag.deleteRepeatGroup(...)` gets called. When this happens, we iterate the all the repeat group instances starting from the position that the one that has been deleted had before deleting it.
+
+**Set of evaluated triggerables**: 
+- those triggered by the repeat group's reference, once per repeat group sibling starting from the position that belonged to the deleted group.
+- those triggered by the references of the children of the deleted repeat group. 
+(see Evaluation of a set of triggerables)
+
+Deleting a repeat group instance will update the group's instance count and the position of all the instances that follow the deleted one. This means that we need to evaluate all triggerables triggered by the repeat group's reference. 
+
+We need to remember that we need to follow an alternative strategy to evaluate the triggerables belonging to repeat groups because the DAG uses genericized trigger references to index sets of triggerables. In this case, calling `Set<QuickTriggerable> alreadyEvaluated = triggerTriggerables(mainInstance, evalContext, repeatGroup.getRef(), new HashSet<>(0))` once per sibling (starting from the deleted group's position) deals with that.
+
+We also have to evaluate any triggerable triggered by children references of the repeat group, which is done by `evaluateChildrenTriggerables(mainInstance, evalContext, repeatGroup, false, alreadyEvaluated)`. Even though this is done inside the loop, the `if` check ensures this is only done once, coinciding with the first iteration (corresponding to the sibling that takes the place of the deleted one). Presumably, this is to have a nice ordering of published events in the `EventNotifier`, although there's no apparent functional requirement for this, and the call could be done outside the loop in `TriggerableDag.deleteRepeatGroup(...)`.
+
+#### d - A complex itemset value is copied to the main instance
```

Comment:
Would be good to add a test form for this.

It's not part of the [spec](https://opendatakit.github.io/xforms-spec/). Forms that use the feature have to be built by hand. In the longer run, should we consider deprecating it and maybe removing that code? It seems that reducing complexity for dynamic selects would be beneficial all around.

This would be useful if, for example, select options were "people" nodes with "phone number", "address", etc child nodes and selecting one "people" node would copy over the entire subtree, right?

Reply:
> It's not part of the spec. Forms that use the feature have to be built by hand

Ahhh, I was suspecting that. Then yes, I also think that we should get rid of it. 

I don't think we could actually deprecate it though, because clients don't actually interact with this feature (there's no public API method that we could tag). The feature automatically kicks in when a complex structure is detected in the values of a select field's itemset.

> This would be useful if, for example, select options were "people" nodes with "phone number", "address", etc child nodes and selecting one "people" node would copy over the entire subtree, right?

Yes, that seems to be the use for this, but I couldn't find any working example of this. After studying the whole DAG's behavior, the code that deals with this is kinda straightforward.

Depending on what we decide to do with this feature, I could try to build a form and add some tests but if we decide to ultimately remove it, then I don't think it's worth the effort.